ruma-events-macros/src/event_content.rs

```diff
@@ -61,65 +147,130 @@ pub fn expand_event_content(input: DeriveInput) -> syn::Result<TokenStream> {
 
+    let redacted_marker_trait = if needs_redacted(input) {
+        let ident = quote::format_ident!("Redacted{}", input.ident);
+        quote! {
+            impl ::ruma_events::EphemeralRoomEventContent for #ident { }
+        }
+    } else {
+        TokenStream::new()
+    };
+
     Ok(quote! {
         #event_content_impl
 
         impl ::ruma_events::EphemeralRoomEventContent for #ident { }
+
+        #redacted_marker_trait
     })
 }
 
 /// Create a `RoomEventContent` implementation for a struct.
-pub fn expand_room_event_content(input: DeriveInput) -> syn::Result<TokenStream> {
+pub fn expand_room_event_content(input: &DeriveInput) -> syn::Result<TokenStream> {
     let ident = input.ident.clone();
     let event_content_impl = expand_event_content(input)?;
 
+    let redacted_marker_trait = if needs_redacted(input) {
+        let ident = quote::format_ident!("Redacted{}", input.ident);
+        quote! {
+            impl ::ruma_events::RoomEventContent for #ident { }
+        }
+    } else {
+        TokenStream::new()
+    };
+
     Ok(quote! {
         #event_content_impl
 
         impl ::ruma_events::RoomEventContent for #ident { }
+
+        #redacted_marker_trait
     })
 }
 
 /// Create a `MessageEventContent` implementation for a struct
-pub fn expand_message_event_content(input: DeriveInput) -> syn::Result<TokenStream> {
+pub fn expand_message_event_content(input: &DeriveInput) -> syn::Result<TokenStream> {
     let ident = input.ident.clone();
     let room_ev_content = expand_room_event_content(input)?;
 
+    let redacted_marker_trait = if needs_redacted(input) {
+        let ident = quote::format_ident!("Redacted{}", &ident);
+        quote! {
+            impl ::ruma_events::MessageEventContent for #ident { }
+        }
+    } else {
+        TokenStream::new()
+    };
+
     Ok(quote! {
         #room_ev_content
 
         impl ::ruma_events::MessageEventContent for #ident { }
+
+        #redacted_marker_trait
     })
 }
 
 /// Create a `StateEventContent` implementation for a struct
-pub fn expand_state_event_content(input: DeriveInput) -> syn::Result<TokenStream> {
+pub fn expand_state_event_content(input: &DeriveInput) -> syn::Result<TokenStream> {
     let ident = input.ident.clone();
     let room_ev_content = expand_room_event_content(input)?;
 
+    let redacted_marker_trait = if needs_redacted(input) {
+        let ident = quote::format_ident!("Redacted{}", input.ident);
+        quote! {
+            impl ::ruma_events::StateEventContent for #ident { }
+        }
+    } else {
+        TokenStream::new()
+    };
+
     Ok(quote! {
         #room_ev_content
 
         impl ::ruma_events::StateEventContent for #ident { }
+
+        #redacted_marker_trait
     })
 }
+
+fn needs_redacted(input: &DeriveInput) -> bool {
+    input
+        .attrs
+        .iter()
+        .flat_map(|a| a.parse_args::<EventMeta>().ok())
+        .find(|a| a == &EventMeta::SkipRedacted || a == &EventMeta::RedactedCustom)
```

Comment:
Ah, you seem to have misunderstood `#[skip_redaction]`. It was supposed to be a *field* attribute, not a struct attribute, and mark fields that should stay when an event is redacted. So this really should just be

```suggestion
        .find(|a| a == &EventMeta::RedactedCustom)
```